cluster-autoscaler/core/scaledown/actuation/delete_in_batch.go

```diff
@@ -0,0 +1,173 @@
+	}
+}
+
+// AddToBucket adds node to delete candidates.
+func (d *NodeBatcherDeleterImpl) AddToBucket(node *apiv1.Node) (string, error) {
+	d.Lock()
+	defer d.Unlock()
+	nodeGroup, err := d.ctx.CloudProvider.NodeGroupForNode(node)
+	if err != nil {
+		return "", err
+	}
+	val, ok := d.deletionsPerNodeGroup[nodeGroup.Id()]
+	if !ok || len(val) == 0 {
+		d.deletionsPerNodeGroup[nodeGroup.Id()] = []*apiv1.Node{node}
+		return nodeGroup.Id(), nil
+	}
+	d.deletionsPerNodeGroup[nodeGroup.Id()] = append(d.deletionsPerNodeGroup[nodeGroup.Id()], node)
+	return "", nil
+}
+
+// CleanUp delete nodes of a given nodeGroup.
+func (d *NodeBatcherDeleterImpl) CleanUp(nodeGroupId string) error {
+	d.Lock()
+	defer d.Unlock()
+	nodes, ok := d.deletionsPerNodeGroup[nodeGroupId]
+	if !ok {
+		return nil
+	}
+	delete(d.deletionsPerNodeGroup, nodeGroupId)
+
+	go func(nodes []*apiv1.Node, nodeGroupId string) {
+		var result status.NodeDeleteResult
+		err := d.deleteNodesFromCloudProvider(nodes)
+		if err != nil {
+			result = status.NodeDeleteResult{ResultType: status.NodeDeleteErrorFailedToDelete, Err: err}
+		} else {
+			result = status.NodeDeleteResult{ResultType: status.NodeDeleteOk}
+		}
+		for _, node := range nodes {
+			if d.nodeDeletionTracker.IsNodeEmpty(node.Name) {
+				d.processEmptyNode(node, nodeGroupId, result)
+			} else if d.nodeDeletionTracker.IsNodeDrained(node.Name) {
+				d.processNonEmptyNode(node, nodeGroupId, result)
+			} else {
+				klog.Errorf("Node %s is not present in nodeDeletionTracker", node.Name)
+			}
+		}
+	}(nodes, nodeGroupId)
+	return nil
+}
+
+// deleteNodeFromCloudProvider removes the given nodes from cloud provider. No extra pre-deletion actions are executed on
+// the Kubernetes side. If successful, the deletion is recorded in CSR, and an event is emitted on the node.
+func (d *NodeBatcherDeleterImpl) deleteNodesFromCloudProvider(nodes []*apiv1.Node) error {
+	d.Lock()
+	defer d.Unlock()
+	nodeGroup, err := d.ctx.CloudProvider.NodeGroupForNode(nodes[0])
+	if err != nil {
+		return errors.NewAutoscalerError(errors.CloudProviderError, "failed to find node group for %s: %v", nodes[0].Name, err)
+	}
+	if nodeGroup == nil || reflect.ValueOf(nodeGroup).IsNil() {
+		return errors.NewAutoscalerError(errors.InternalError, "picked node that doesn't belong to a node group: %s", nodes[0].Name)
+	}
+	if err = nodeGroup.DeleteNodes(nodes); err != nil {
+		return errors.NewAutoscalerError(errors.CloudProviderError, "failed to delete %s: %v", nodes[0].Name, err)
+	}
+	for _, node := range nodes {
+		d.ctx.Recorder.Eventf(node, apiv1.EventTypeNormal, "ScaleDown", "nodes removed by cluster autoscaler")
+		d.clusterState.RegisterScaleDown(&clusterstate.ScaleDownRequest{
+			NodeGroup:          nodeGroup,
+			NodeName:           node.Name,
+			Time:               time.Now(),
+			ExpectedDeleteTime: time.Now().Add(MaxCloudProviderNodeDeletionTime),
+		})
+		drain := d.nodeDeletionTracker.IsNodeDrained(node.Name)
```

Comment:
Hmm, do you think it's worth it to jump through the deletion tracker for this? We could just pass and store the drain variable along with the node in this structure.

Reply:
We already store the data if node was drain or not in tracker, why store it in another place?

Reply:
Mostly because it's a simple property that we have available when we add node to the batcher anyway, so we could avoid changing the tracker and introducing a function call. But that's not a very meaningful difference, so probably down to personal preferences. An argument could also be made for reducing lock contention in the tracker, but it'll probably be fine in practice (and we can always change it if we see a scalability regression). Just wanted to flag this as an option, the current approach also LGTM.

Reply:
You're right. I added a map with drained nodes to the batcher directly and tested both versions, the versions with drained nodes map perform better (for 1000 empty nodes cluster, scale down was faster for 1 minute).